### ISSUE TITLE
KroneckerMultiTaskGP Not performing as expected

### ISSUE THREAD

**Issue by karabelaselias** [pytorch/botorch#1008]
# KroneckerMultiTaskGP Not performing as expected

Hi there, as I'm quite new to the field of Gaussian Processes I wasn't sure where to post this so I decided for issuing a bug report.
What I'm trying to achieve is a fit of a multi-input / multi-output GP. I basically have Input Parameters x_0,....x_d and m output features y_0, .... y_m and I want to fit the vector-valued function that maps x -> y.

After studying most of the tutorials here on the site I came to the understanding, that the KroneckerMultiTaskGP seems to be what I want. So I had now samples  x_train  (nsamples x d) and observed y's (nsamples x m).
I was able to train the GP and sample the posterior but somehow the result doesn't look very confident.

![foo](https://user-images.githubusercontent.com/9953640/144417757-806533b7-6c4d-475b-ae07-f1fde583a7f3.png)

So what I observe is, that per row I seem to have the same mean and confidence interval, did I understand the KroneckerMultiTaskGP wrong?

I'll try to add my code and the dataset I used that reproduces it.

## To reproduce

** Code snippet to reproduce **
```python
import torch
import os
import time
import numpy as np
import pandas as pd
from scipy import stats

from botorch.models import KroneckerMultiTaskGP
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.optim.fit import fit_gpytorch_torch

# load datasets remove outliers, scale, and standarize them
X = np.loadtxt("X.txt", dtype=float) # yields a 87 x 4 tensor (87 samples for 4 parameters from a LHD)
Y = np.loadtxt("Y.txt", dtype=float) # yields a 87 x 14 tensor (14 output variables observed / calculated)

dfY = pd.DataFrame(data=Y)
dfX = pd.DataFrame(data=X)
locs = (np.abs(stats.zscore(dfY, ddof=1)) < 3.0).all(axis=1)
dfY = dfY[locs]
dfX = dfX[locs]

X = dfX.to_numpy()
Y = dfY.to_numpy()
    
a = np.min(X, axis=0)
b = np.max(X, axis=0)
    
Xscal = (X-a) / (b-a)
    
Ymean = np.mean(Y, axis=0)
Ystd = np.std(Y, axis=0)
Yscaled = (Y - Ymean) / Ystd

Xtorch = torch.from_numpy(Xscal).float()
Ytorch = torch.from_numpy(Yscaled).float()

torch.random.manual_seed(2010)
if torch.cuda.is_available():
    torch.cuda.set_device("cuda:0")
tkwargs = {
    "dtype": torch.double,
    "device": torch.device("cuda:0" if torch.cuda.is_available() else "cpu"),
}
gp = KroneckerMultiTaskGP(Xtorch, Ytorch)
gp_mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_torch(gp_mll, options={"maxiter": 3000, "lr": 0.001, "disp": False})

#plot the fit 
from matplotlib import pyplot as plt
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.stack([
            torch.linspace(0, 1, Xtorch.shape[0]),
            torch.linspace(0, 1, Xtorch.shape[0]),
            torch.linspace(0, 1, Xtorch.shape[0]),
            torch.linspace(0, 1, Xtorch.shape[0]),
        ]).T
    postf = gp.posterior(test_x, observation_noise=True)
    mean = postf.mean
    var = postf.variance
    lower = mean - 2 * var.sqrt()
    upper = mean + 2 * var.sqrt()
# Initialize plots
f, axs = plt.subplots(Ytorch.shape[1], Xtorch.shape[1], figsize=(12, 24))
axs = axs.flatten()
for i in range(Ytorch.shape[1]):
    for j in range(Xtorch.shape[1]):
        axs[Xtorch.shape[1]*i+j].plot(Xtorch.detach().numpy()[:,j], Ytorch[:, i], 'k*')
        axs[Xtorch.shape[1]*i+j].plot(test_x.numpy()[:,j], mean[:, i].numpy(), 'b')
        axs[Xtorch.shape[1]*i+j].fill_between(test_x.numpy()[:,j], lower[:, i].numpy(), upper[:, i].numpy(), alpha=0.5)
# this produces the figure above.

```

## Expected Behavior

I was expecting the GP to fit the first row better, since there is a clear linear behavior visible, but somehow there are some features that are fitted well and others are just ignored or not fitted.

## System information

**Please complete the following information:**
- BoTorch Version 0.5.1
- GPyTorch Version 1.5.1
- PyTorch Version 1.10.1
- Ubuntu 20.04 LTS

## Additional context

I hope someone can point me in the right direction to look, and sorry If I couldn't find the appropriate point in the issue tracker to add my request.

[X.txt](https://github.com/pytorch/botorch/files/7641734/X.RV.txt)
[Y.txt](https://github.com/pytorch/botorch/files/7641735/Y.RV.txt)

**Comment by Balandat** [comment]
Hmm so I guess first of all I'm a bit confused as to what exactly you're plotting what you are plotting. Each row seems to be an output, but then each column is a single-dimensional slice in the respective coordinate. But of course by doing that you're not really taking into account the value of the other features. So I don't think this plot is all that useful for understanding model fit. Instead, I would suggest you look at a cross-validation plot of the outcomes to see how well the model fits. 

A couple of other things (though my main concern is how I think you might be misinterpreting the plots):
- Are your observations noisy? If not, there is no value in using a multi-task GP here (see https://github.com/pytorch/botorch/discussions/1003 for relevant discussion)
- Have you tried to fit independent GPs?
- Using `fit_gpytorch_torch` with a only 300 iterations seems problematic here - There are a lot of tasks and so a lot of hyperparameters to fit, and I doubt the optimizer will converge in 300 steps here. Have you tried `fit_gpytorch_scipy` here?

**Comment by karabelaselias** [comment]
@Balandat Thanks for the response!

So I hope I can clarify a few things

1. My understanding was, that I can fit a multi-task GP to get a regression of y = GP(x) so that's why I plotted then the results as realizations of the posterior sampled at some test points to check the quality of regression for each y
2. I already fitted independent GPs which works ok, I just wanted to see if I can gain anything by using the multitask gp because I'm not quite sure if the assumption of independence will hold (although I found some convincing arguments in the Book of Gramcy why you can still get away with it) and I suspect some of the parameters to be correlated.
3. I assume the observation parameters to have some noise attached ( the input parameters for sure)
3. My ultimate goal is to plug the surrogate model created this way into a global sensitivity analysis (SALib)

**Comment by wjmaddox** [comment]
I just took a look at your code and a couple of things:

- A learning rate of 0.001 is probably too small; a learning rate of 0.01 seems to converge faster and to a similar loss. The higher learning rate model tends to perform better in my understanding, but here there's probably marginal utility in `fit_gpytorch_scipy` due to the overparameterization of the model.
- I'm getting better looking fits on your data by setting the initial values of the likelihood to be smaller, via `likelihood.noise = torch.ones(1) * 0.1` and `likelihood.task_noises = torch.ones(14) * 0.1`. But, there's still a bit of overconfidence rather than underconfidence.
- I think there's something a bit funky about your testing plot code as you're showing the posterior over [0,1]^4 where each example is [0.1, 0.1, 0.1, 0.1]. Using a real held-out test set here gives pretty solid looking predictions (I used a random split of 50 of your 87 points to produce the following plot).

![image](https://user-images.githubusercontent.com/13008542/144870020-f3257dc5-478d-4fab-9f57-d4144dbf3379.png)

**Comment by karabelaselias** [comment]
@wjmaddox Thanks for the reply, I wasn't sure how to plot the whole thing, I followed some of the SingleGP tutorials of gpytorch to see the quality of the fit but this clearly seems to produce wrong plots. Could you tell me how you did this plot?

**Comment by wjmaddox** [comment]
Sure, here's my code:

```python
inds = torch.randint(Xtorch.shape[0], torch.Size((Xtorch.shape[0],)))

train_x = Xtorch[inds[:50],]
train_y = Ytorch[inds[:50],]

test_x = Xtorch[inds[50:],]
test_y = Ytorch[inds[50:],]

gp = KroneckerMultiTaskGP(
    train_x, train_y, 
)
gp_mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_torch(gp_mll, options={"maxiter": 3000, "lr": 0.01, "disp": True})

with torch.no_grad():
    postf = gp.posterior(test_x, observation_noise=True)

fig, ax = plt.subplots(2, 7, figsize = (40, 10))

ax = ax.reshape(-1)

for i in range(14):
    ax[i].errorbar(test_y[:, i], postf.mean[:, i], yerr=2 * postf.variance[:, i]**0.5, ls = "none",
                  marker = ".", markersize = 20)
```

If you want to create a profile based plot like you were showing originally, you'll have to do something where you modify only one feature at a time like this:

```python
profile_x = torch.ones(50, 4) / 2 # mean should be at about 0.5^4 

posteriors = []
with torch.no_grad():
    for i in range(4):
        profile_x[:, i] = torch.linspace(0, 1, 50)
        postf = gp.posterior(profile_x, observation_noise = True)
        posteriors.append(postf)
        profile_x[:, i] = torch.ones(50) / 2 # put it back

f, axs = plt.subplots(14, 4, figsize = (12, 24))
axs = axs.T

for i in range(4):
    for j in range(14):
        axs[i,j].plot(Xtorch[:,i], Ytorch[:, j], 'k*')
        axs[i,j].plot(torch.linspace(0, 1, 50), posteriors[i].mean[:, j], "b")
        axs[i,j].fill_between(
            torch.linspace(0, 1, 50), 
            posteriors[i].mean[:, j] - 2 * posteriors[i].variance[:, j].sqrt(),
            posteriors[i].mean[:, j] + 2 * posteriors[i].variance[:, j].sqrt(),
            alpha = 0.3
        )
```

which looks pretty good if potentially underfit:
![image](https://user-images.githubusercontent.com/13008542/144874091-a8ddb97a-4647-49fb-8a6e-950e4dd44f5f.png)

**Comment by Balandat** [comment]
@karabelaselias hopefully this discussion was helpful and resolved your issues, please feel free to reopen this issue if needed.
